cedar-policy-core/src/ast/restricted_expr.rs

```diff
@@ -309,16 +320,20 @@ impl<'a> Hash for RestrictedExprShapeOnly<'a> {
     }
 }
 
-/// Errors generated in the restricted_expr module
+/// Errors related to restricted expressions
 #[derive(Debug, Clone, PartialEq, Error)]
 pub enum RestrictedExprError {
-    /// A "restricted" expression contained a feature that is not allowed
-    /// in "restricted" expressions. The `feature` is just a string description
-    /// of the feature that is not allowed.
-    #[error("not allowed to use {feature} in a restricted expression")]
+    /// An expression was expected to be a "restricted" expression, but contained
+    /// a feature that is not allowed in restricted expressions. The `feature`
+    /// argument is a string description of the feature that is not allowed.
+    /// The `expr` argument is the expression that uses the disallowed feature.
+    /// Note that it is potentially a sub-expression of a larger expression.
+    #[error("not allowed to use {feature} in a restricted expression: {expr}")]
     InvalidRestrictedExpression {
         /// what disallowed feature appeared in the expression
         feature: String,
+        /// the (sub-)expression that uses the disallowed feature
+        expr: Expr,
```

Comment:
while we're changing the fields of this variant anyway, want to change `String` to `SmolStr`?

Reply:
Sure. Done in the latest commit.